draco/synthesis/mockcatalog.py

```diff
@@ -0,0 +1,1045 @@
+                # Get global index of z bin containing source, and central z
+                global_z_index = zi + self.lo
+                z_value = z_global["centre"][global_z_index]
+
+                # If desired, add random offset within z bin
+                if not self.z_at_channel_centers:
+                    z_value += z_global["width"][global_z_index] * rz[zi][si]
+
+                # If desired, add Gaussian z error
+                if self.sigma_z is not None:
+                    err = rzerr[zi][si] * self.sigma_z
+                    z_value += err
+                    mock_zerrs[source_count] = err
+                elif self.sigma_z_over_1plusz is not None:
+                    err = rzerr[zi][si] * self.sigma_z_over_1plusz * (1 + z_value)
+                    z_value += err
+                    mock_zerrs[source_count] = err
+                else:
+                    mock_zerrs[source_count] = 0
+
+                # Populate local arrays of source redshift, RA, dec,
+                # adding random angular offsets from pixel centers if desired
+                mock_zs[source_count] = z_value
+                mock_ra[source_count] = RAbase
+                mock_dec[source_count] = decbase
+                if not self.srcs_at_pixel_centers:
+                    mock_ra[source_count] += ang_size * rtheta[zi][si]
+                    mock_dec[source_count] += ang_size * rphi[zi][si]
+
+                source_count += 1
+
+        # Define arrays to hold full source catalog
+        mock_zs_full = np.empty(self.nsources, dtype=mock_zs.dtype)
+        mock_zerrs_full = np.empty(self.nsources, dtype=mock_zs.dtype)
+        mock_ra_full = np.empty(self.nsources, dtype=mock_ra.dtype)
+        mock_dec_full = np.empty(self.nsources, dtype=mock_dec.dtype)
+
+        # Tuple (not list!) of number of sources in each rank
+        # Note: the counts and displacement arguments of Allgatherv are tuples!
+        nsource_tuple = tuple(self.comm_.allgather(nsource_rank))
+        # Tuple (not list!) of displacements of each rank array in full array
+        dspls = tuple(np.insert(arr=np.cumsum(nsource_tuple)[:-1], obj=0, values=0.0))
+        # Gather redshifts
+        recvbuf = [mock_zs_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_zs, len(mock_zs)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
+        # Gather redshift errors
+        recvbuf = [mock_zerrs_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_zerrs, len(mock_zs)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
+        # Gather theta
+        recvbuf = [mock_ra_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_ra, len(mock_ra)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
+        # Gather phi
+        recvbuf = [mock_dec_full, nsource_tuple, dspls, MPI.DOUBLE]
+        sendbuf = [mock_dec, len(mock_dec)]
+        self.comm_.Allgatherv(sendbuf, recvbuf)
```

Comment:
Pro-level use of `Allgatherv`. Kudos.

Reply:
All kudos goes to @fandinomat, who wrote all the nontrivial parts!